src/textioc/textioc.asm

```diff
@@ -0,0 +1,3970 @@
+; Returns:
+;   HL = approximated pointer to next line
+;   A = 0 if function found a NULL
+; Destroys:
+;   All working registers
+
+; Save pointer to start of line
+	push	hl
+.loop:
+	pop	de
+	push	hl
+	push	de
+	call	textio_GetLineWidth
+	ex	de,hl
+	pop	bc				; BC = pointer to start of line
+	pop	hl				; HL = pointer to current character
+	push	bc
+	push	hl
+	call	textio_GetTextWindowWidth
+	or	a,a
+	sbc	hl,de
+	pop	hl				; HL = pointer to current character
+	pop	de				; DE = pointer to start of line
+	ld	a,1
+	ret	c
+
+	call	textio_GetNewlineCode
+	cp	a,(hl)
+	ret	z
+
+	ld	a,null
+	cp	a,(hl)
+	ret	z
+
+	ld	bc,0
+	ld	c,(hl)
+	push	de
+	push	hl
+	push	bc
+	call	textio_GetCharWidth
+
+; Add character width to line width and increment HL
+	pop	bc
+	ex	de,hl
+	ld	hl,(_LineWidth)
+	add	hl,de
+	ld	(_LineWidth),hl
+	pop	hl
+	inc	hl
+	jr	.loop
+
+
+;-------------------------------------------------------------
+util.breakString:
+; Arguments:
+;   HL = pointer to start of string
+; Returns:
+;   Pointer to the remainder of the string
+; Destroys:
+;   A, BC, DE, and HL
+
+; General register values:
+; BC = number of chars in string
+; DE = width of string
+; HL = pointer to current char in string being processed
+
+
+; Set B to zero
+	ld	b,0
+	
+.loop:
+; Determine if the current char is a NULL
+	push	hl
+	ld	a,null
+	cp	a,(hl)
+	pop	hl
+	ret	z
+; Determine if the current char is a tab or glyph
+; If printing RMF, spaces and tabs are appended to the start of
+; the string
+	ld	a,tab
+	cp	a,(hl)
+	jr	nz,.getGlyphWidth
+	push	de
+	push	hl
+	or	a,a
+	sbc	hl,hl
+	ld	l,tab
+	push	hl
+	call	textio_GetCharWidth
+	pop	hl
+	pop	hl
+	pop	de
+	jr	.addToStrWidth
+.getGlyphWidth:
+	push	hl
+	or	a,a
+	sbc	hl,hl
+	ex	de,hl
+	pop	hl
+	ld	e,(hl)
+	push	hl
+	push	de
+	call	textio_GetCharWidth
+	pop	de
+	pop	hl
+.addToStrWidth:
+; Add tab or glyph width to string width (DE)
+	push	hl
+	push	af
+	or	a,a
+	sbc	hl,hl
+	pop	af
+	ld	de,(_LineWidth)
+	ld	l,a
+	add	hl,de
+	ex	de,hl
+; If line width + fragment width < window width, jump to loop
+	call	textio_GetTextWindowWidth
+	or	a,a
+	sbc	hl,de
+	pop	hl
+	ret	c
+; Since char was successfully added, add its width to the line width
+	ld	(_LineWidth),de
+; Increment current char pointer and current char number
+	inc	hl
+	inc	b
+	jr	.loop
+
+	
+;-------------------------------------------------------------
+util.startNewLine:
+; Arguments:
+;   HL = pointer to current character being processed in the line
+; Returns:
+;   Adds one to _CurrLineNum
+;   Sets BC equal to HL
+;   Zeros the _LineWidth
+
+; Increment the current line number
+	ld	a,(_CurrLineNum)
+	inc	a
+	ld	(_CurrLineNum),a
+; Set the current line pointer to the current char pointer
+	ex	de,hl
+	xor	a,a
+	sbc	hl,hl
+	ld	(_LineWidth),hl
+	ex	de,hl
+	push	hl
+	pop	bc
+	ret
+	
+
+
+
+;-------------------------------------------------------------
+; Internal Library Functions
+;-------------------------------------------------------------
+
+
+util.CallHL:
```

Comment:
this should probably check that hl != 0 rather than just blindly jumping if the user is dumb

Reply:
Agreed. I will fix that and update the documentation tomorrow.